nilearn/input_data/base_masker.py

```diff
@@ -82,136 +92,68 @@ def filter_and_mask(imgs, mask_img_,
-                        low_pass=parameters['low_pass'],
-                        high_pass=parameters['high_pass'],
-                        t_r=parameters['t_r'],
-                        detrend=parameters['detrend'],
-                        standardize=parameters['standardize']
-            )
-
-    # For _later_: missing value removal or imputing of missing data
-    # (i.e. we want to get rid of NaNs, if smoothing must be done
-    # earlier)
-    # Optionally: 'doctor_nan', remove voxels with NaNs, other option
-    # for later: some form of imputation
-
-    return data, imgs.get_affine()
+    if verbose > 0:
+        print("[%s] Cleaning extracted signals" % class_name)
+    sessions = parameters.get('sessions', None)
+    region_signals = cache(
+        signal.clean, memory=memory, func_memory_level=2,
+        memory_level=memory_level)(
+            region_signals,
+            detrend=parameters['detrend'],
+            standardize=parameters['standardize'],
+            t_r=parameters['t_r'],
+            low_pass=parameters['low_pass'],
+            high_pass=parameters['high_pass'],
+            confounds=confounds,
+            sessions=sessions)
+
+    return region_signals, aux
 
 
 class BaseMasker(BaseEstimator, TransformerMixin, CacheMixin):
     """Base class for NiftiMaskers
     """
 
-    def transform_single_imgs(self, imgs, confounds=None, copy=True,
-                              sample_mask=None):
-        if not hasattr(self, 'mask_img_'):
-            raise ValueError('It seems that %s has not been fitted. '
-                             'You must call fit() before calling transform().'
-                             % self.__class__.__name__)
-        params = get_params(self.__class__, self)
-        # Remove the mask-computing params: they are not useful and will
-        # just invalid the cache for no good reason
-        for name in ('mask_img', 'mask_args'):
-            params.pop(name, None)
-        data, _ = self._cache(filter_and_mask,
-                              ignore=['verbose', 'memory', 'copy'])(
-                                  imgs, self.mask_img_,
-                                  params,
-                                  memory_level=self.memory_level,
-                                  memory=self.memory,
-                                  verbose=self.verbose,
-                                  confounds=confounds,
-                                  copy=copy,
-                                  sample_mask=sample_mask
-        )
-        return data
-
-    def transform_imgs(self, imgs_list, confounds=None, copy=True, n_jobs=1):
-        ''' Prepare multi subject data in parallel
+    @abc.abstractmethod
+    def transform_single_imgs(self, imgs, confounds=None, copy=True):
+        """Extract signals from a single 4D niimg.
 
         Parameters
-        ----------
-
-        imgs_list: list of Niimg-like objects
+        ==========
+        imgs: Niimg-like object
```

Comment:
Yes, it's written below. I think that 3D niimg work too since it is converted to 4D with a single scan as last dimension.